server/src/main/java/org/opensearch/transport/InboundHandler.java

```diff
@@ -246,32 +247,46 @@ private void messageReceived(TcpChannel channel, InboundMessage message, long st
                     assertRemoteVersion(stream, header.getVersion());
                     final RequestHandlerRegistry<T> reg = requestHandlers.getHandler(action);
                     assert reg != null;
-                    final T request = reg.newRequest(stream);
-                    request.remoteAddress(new TransportAddress(channel.getRemoteAddress()));
-                    // in case we throw an exception, i.e. when the limit is hit, we don't want to verify
-                    final int nextByte = stream.read();
-                    // calling read() is useful to make sure the message is fully read, even if there some kind of EOS marker
-                    if (nextByte != -1) {
+
+                    try {
+                        final T request = reg.newRequest(stream);
```

Comment:
> The EOFException will come from this line, right?

Correct

> I'd prefer to keep the EOFException try/catch more tightly bounded in order to make things easier to follow with less indentation. What do you think?

I thought about that as well, the issue is that `request` is used at the end of the code block as well

```
...

if (ThreadPool.Names.SAME.equals(executor)) {
    try {
        reg.processMessageReceived(request, transportChannel);
    } catch (Exception e) {
        sendErrorResponse(reg.getAction(), transportChannel, e);
    }
} else {
    threadPool.executor(executor).execute(new RequestHandler<>(reg, request, transportChannel));
}
```
So the alternative would introduce `nullable` request instance, no sure it is better (but readability would suffer I think). With the `IllegalStateException` check wrapped into dedicated function it may look better actually

Reply:
Since the catch block never returns you can do this:

```java
final T request;
try {
    request = reg.newRequest(stream);
} catch (EOFException e) {
    // Another favor of (de)serialization issues is when stream contains less bytes than
    // the request handler needs to deserialize the payload.
    throw new IllegalStateException(
        "Message fully read (request) but more data is expected for requestId ["
            + requestId
            + "], action ["
            + action
            + "]; resetting",
        e
    );
}
request.remoteAddress(new TransportAddress(channel.getRemoteAddress()));
checkStreamIsFullyConsumed(requestId, action, stream);

final String executor = reg.getExecutor();
if (ThreadPool.Names.SAME.equals(executor)) {
    try {
        reg.processMessageReceived(request, transportChannel);
    } catch (Exception e) {
        sendErrorResponse(reg.getAction(), transportChannel, e);
    }
} else {
    threadPool.executor(executor).execute(new RequestHandler<>(reg, request, transportChannel));
}
```

Reply:
That the part I don't like to be fair:

```
final T request;

try {
    request = reg.newRequest(stream);
}

```

I would try with the function instead, just a moment

Reply:
@andrross done, thank you!